### ISSUE TITLE
Siamese architecture, multi-label output

### ISSUE THREAD

**Issue by sanjeevmk** [fchollet/keras#2675]
I've a data set 4 examples each of 50 inputs ,each input of dimension 3 (that is, data.shape=(4,50,3) )

I've implemented a siamese network exactly as given in this example:
https://github.com/fchollet/keras/blob/master/examples/mnist_siamese_graph.py

To draw correspondence, the way the above example has N pairs of inputs, I have 4 examples of 50 inputs.
First difference is that instead of a pair of input features, I have 50 input features each of 3 dimensions.
Each of these 50 inputs (of each example) go through a shared network exactly as in the example above, and the final output of the shared network is of 6 dimensions (that is, the last layer of the shared network has 6 neurons). 
So after these 50 inputs pass through the shared network, I'm expecting 50 6-dimensional vectors as output of the shared network.
After the shared layers, I want to add a Lambda layer that operates on these 50 vectors. In the example linked above, the final Lambda layer takes as input 2 vectors and computes the euclidean distance. Instead, I want to take in these 50 vectors (6-D) and sum them along the columns. For example, if the inputs to the final layer are [1,1,1,1,1,1] and [2,2,2,2,2,2], I want the output to be [3,3,3,3,3,3].

With that in mind, I wrote the following:           

```
    def summation(vectors):
        return K.sum(vectors,axis=1)

    def final_output_shape(self,shapes):
        return shapes[0]

    def create_base_network():
        seq = Sequential()
        seq.add( Dense(6,input_shape=(3,),activation='relu') )
        return seq

    inputs = [Input(shape=(3,)) for i in range(50)]
    base_network = create_base_network(input_dim)
    processed_inputs = [base_network(i) for i in inputs]
    label_output = Lambda(summation,output_shape=(6,) )(processed_inputs)
    model = Model(input=inputs,output=label_output)
    model.compile(loss='mse',optimizer='RMSprop')


```

This compiles fine. For the last layer, if I do `K.sum(vectors)`, then it works fine, but that gives me a single scalar output after the final Lambda layer (which I don't want). But if I do `K.sum(vectors,axis=1)`, the compilation works , but when I call `fit`, I get a dimension mismatch error at the _last layer_ .
Printing `len(vectors)` in the `summation` function prints 50, that means I am getting 50 tensor variables as input to the Lambda layer, as expected, but I'm unable to sum along axis 1 for some reason..

What am I doing wrong? Am I making some wrong assumptions here.  

The specific error I'm getting is like this:  
ValueError: Input dimension mis-match. (input[0].shape[0] = 50, input[1].shape[0] = 4)
Apply node that caused the error: Elemwise{Sub}[(0, 0)](Sum{axis=[1], acc_dtype=float64}.0, lambda_1_target)
Toposort index: 429
Inputs types: [TensorType(float32, matrix), TensorType(float32, matrix)]
Inputs shapes: [(50, 6), (4, 6)]
Inputs strides: [(24, 4), (24, 4)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[InplaceDimShuffle{0,x,1}(Elemwise{Sub}[(0, 0)].0), Elemwise{Sqr}[(0, 0)](Elemwise{Sub}[%280, 0%29].0)]]

In the above message I don't know why those input shapes [(50,6),(4,6)] are there.. Note that the '4' here is the number of examples (The '4' in input[1].shape[0]=4 is the same)

**Comment by braingineer** [comment]
I'm guessing this is theano. 

The square in the output client seems like it isn't in that code at all. your loss is mean squared error. so I'm guessing the problem also involves your targets.  what is the shape of the data you're feeding?

Some tips on debugging:

you can run your code as:
```python
THEANO_FLAGS='optimizer=None,exception_verbosity=high' python file.py 
```
It'll give you a pinpoint on the code that creates the error. 

You can also use theano.printing.Print() to print intermediate states.

Finally, and my favorite method of debugging, you can either make the keras graph and the construct a theano function from an intermediate point, or you can just step through the operations yourself in an ipython notebook or something. 

So like, after you compile
```python
F = theano.function(inputs+[K.learning_phase()], processed_outputs+[label_output])
outs = F(x,y)
for out in outs:
    print(out.shape)
```
or something

**Comment by sanjeevmk** [comment]
@braingineer Thanks for the inputs, I'll try these debugging methods..

My input shapes are as follows:  
I first create a feature array `features` whose shape is (4,50,3) --> 4 examples, 50 inputs, each of 3D  
The target shape is (4,6) --> 4 examples, each has a 6D output.  

I call my `fit` this way, which is exactly as in the mnist siamese graph example:  
`model.fit(  [features[:,i] for i in range(50)] , labels , epochs=nb_epochs)`  

When called this way, I'm basically passing 50 arrays, each of shape (4,3)  (4 examples, 3D input features)  . `labels` are passed as is, with shape (4,6).  

Also, the error message also had this:  
Backtrace when the node is created(use Theano flag traceback.limit=N to make it longer):
  File "siamesebasicmodel.py", line 145, in <module>
    nn.initialize(num_inputs,input_dim,output_dim)
  File "/media/sanjeevmk/Progammer/FunctionalityNetwork/src/KerasSiamese.py", line 35, in initialize
    self.model.compile(loss='mse',optimizer='sgd')
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.0.2-py2.7.egg/keras/engine/training.py", line 588, in compile
    sample_weight, mask)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.0.2-py2.7.egg/keras/engine/training.py", line 311, in weighted
    score_array = fn(y_true, y_pred)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.0.2-py2.7.egg/keras/objectives.py", line 7, in mean_squared_error
    return K.mean(K.square(y_pred - y_true), axis=-1)

When I run using the 'exception verbosity' flag as high, I get a huge stack trace of the apply node, which is hard to make sense of..

**Comment by sanjeevmk** [comment]
Hi,
I did some debugging, and I'm not sure but it looks like a problem with `K.sum` 

First , I wrote this:  
```
	def get_layer_output(self,X,n):
		get_nth_layer_output = K.function([model.layers[i].input for i in range(50)],model.layers[n].output)
		layer_output = get_nth_layer_output(X)
		return layer_output
```
, 
It takes as input a feature and a layer number and gives the output of that layer after processing the feature.  

Again, the final Lambda layer is supposed to receive 50 6-D vectors, which it does receive. Now let this Lambda layer be made to compute `K.sum(vectors,axis=1)`. In this case, if I call `out = get_layer_output(one_example_feature,-1)` and print the shape of `out`, it prints (50,6). This is wrong..after `K.sum(vectors,axis=1)` the output should be of shape (1,6), not (50,6). `K.sum` is supposed to take the incoming (50,6) and sum across columns to give a (1,6) final output shape.  

Next, lets make the Lambda layer just return one of the incoming vectors by using `return vectors[0]` . So the lambda layer receives 50 vectors and we just return the first vector back. In this case, if I call `out=get_layer_output(one_example_feature,-1)`, then I get the output shape as (1,6) --> exactly as required.  So when the lambda layer returns `vector[0]` everything compiles and runs fine , which means my input features are of the correct shape (obviously returning the first vector is useless for me, but it serves to highlight the issue).  

To summarize, `K.sum(vectors,axis=1)` should output a shape of (1,6) after receiving (50,6) input, but it doesn't..

**Comment by sanjeevmk** [comment]
I now changed the Lambda layer to do the sum over 50 vectors without using `K.sum`, and everything works perfectly.  

Earlier , the Lambda layer computed the sum like below and it didn't work:  
```
	def summation(vectors):
          return K.sum(vectors,axis=1)

```

Now, I changed the summation function to:  

```
	def summation(vectors):
		v = vectors[0]
		for _v in vectors[1:]:
			v = v+_v
		return v

```

Now, everything works as expected, the output shape is now (1,6) instead of (50,6).. I think the problem is with `K.sum`. I'm using the latest version of Keras. I downloaded the zip from github and ran 'python setup.py install'

**Comment by braingineer** [comment]
Hi @sanjeevmk 

it's getting a bit late over here but I'll try to respond to a couple things.

> When I run using the 'exception verbosity' flag as high, I get a huge stack trace of the apply node, which is hard to make sense of.. 

Haha. ya. it's pretty intimidating, but the bit that's useful you caught:

>return K.mean(K.square(y_pred - y_true), axis=-1) 

this means there's a mis-match between the predicted and the true.  my guess is a shape is misbehaving (it's the usual bug). 

I actually made a PR earlier to address this issue, I think..

the lambda shape parameter doesn't specify the FULL shape.  it specifies the X part of the `(nb_samples,)+X` shape. So, if you pass in `(6,)` to the output_shape parameter, Lambda things you mean `(nb_samples,6)` where nb_samples is taken from the input. 

This explains where the 50 is coming from.  

The PR request I put in is to have this either be better explained or be made more consistent.  Because, if you pass in a function instead of the tuple, (arbitrarily, try `lambda *args: (6,)`), then it will think that the function specifies the FULL shape, rather than everything beyond dim0.  

I hope this is illuminating.  I highly suggest checking out the Lambda function code in core.py.

**Comment by sanjeevmk** [comment]
@braingineer Sorry yeah, other than the `K.sum` thing I changed the output shape to (1,6) instead of (6,)..but even then `K.sum` didn't work..I'm having to do the sum over vectors using `+` as explained in the last post. Maybe the problem is on my end, but that doesn't explain why replacing `K.sum` with `+` over vectors worked..

**Comment by braingineer** [comment]
I think you might have misunderstood or I am currently misunderstanding you. 

Having the code as 

```python
Lambda(some_func, output_shape=(1,6))
```
Means keras will think you are outputting: `(nb_samples, 1, 6)`

**Comment by sanjeevmk** [comment]
Okay, tried that too..didn't work. The function is receiving 50 vectors, and I'm able to iterate over them and sum them up "manually" . The results are as I expect , so things are working. But yeah, I'm not sure why things are breaking when I do it with `K.sum` though..My problem is solved, but had to work around the library..

**Comment by braingineer** [comment]
> tried that too.. didn't work.

You just tried something that should have resulted in completely different behavior.  I'm a little skeptical as to your flippancy.  

in either case, the statement "had to work around the library" is a bit dishonest in how it assigns blame.  Looking at the code would have illuminated the shape problem.  

Also, are you trying any of your code in theano along side your implementation?

> To summarize, K.sum(vectors,axis=1) should output a shape of (1,6) after receiving (50,6) input, but it doesn't..

That's not true.  The first axis is the 6.   Summing over the first axis for (50,6) should result in (50,1). 

Try it in numpy even. 

```python
In [1]: import numpy as np

In [2]: a = np.ones((50,6))

In [3]: np.sum(a, axis=1).shape
Out[3]: (50,)

In [4]: np.sum(a, axis=0).shape
Out[4]: (6,)

In [5]: np.sum(a, axis=0,keepdims=True).shape
Out[5]: (1, 6)

In [6]: np.sum(a, axis=1,keepdims=True).shape
Out[6]: (50, 1)

```

**Comment by sanjeevmk** [comment]
I'm sorry I got confused about the `axis=1` thing, you're right, I should use `axis=0`. I'm not being flippant, I was only trying to highlight that the problem was not in the input/target shapes (since you asked me to investigate that part in your first reply). 
I found out the problem, it was a silly mistake on my end.  

The problem was this:  
I was doing `K.sum(vectors)`, without specifying an axis parameter. I thought `axis=0` was the default, which was wrong, the default was None. I changed my lambda function to return `K.sum(vectors,axis=0)` , specifying axis=0 this time, and everything fell into place now.  
I'm still specifying my output shape as (1,6) and it works..

It seems that was the only problem throughout.. Thanks for your help @braingineer
